api/envoy/extensions/http/stateful_session/header/v3/header.proto

```diff
@@ -28,15 +28,61 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 //
 // .. code-block:: none
 //
-//     session-header: "MS4yLjMuNDo4MA=="
+//     session-header: "MS4yLjMuNDo4MA==" # base64("1.2.3.4:80")
 //
 // When processing the upstream response, if ``1.2.3.4:80`` is indeed the final choice the extension
 // does nothing. If ``1.2.3.4:80`` is not the final choice, the new selected host will be set to
 // response headers (via the ``session-header`` response header).
 //
 // [#extension: envoy.http.stateful_session.header]
 message HeaderBasedSessionState {
-  // The name that will be used to obtain header value from downstream HTTP request or generate
-  // new header for downstream.
+  // The override mode config for the header based session state extension.
+  message OverrideModeConfig {
+  }
+
+  // The packages mode config for the header based session state extension.
+  message PackagesModeConfig {
+  }
+
+  // The name that will be used to obtain header value from downstream HTTP request and generate
+  // new response header for downstream. The Envoy assume the upstream server or the Envoy itself
+  // will generate the ``name`` field specified header in the response. The header value will
+  // contains the necessary information to do the host selection. And Envoy assume the clients
+  // will send the same header back in the subsequent requests if they receive the header and want
+  // to use the same session.
   string name = 1 [(validate.rules).string = {min_len: 1}];
+
+  // Set this to enable override mode. In this mode, if the upstream host address is different with
+  // the target host address parsed from the ``name`` field specified header in the request, or if
+  // the ``name`` field specified header is not present in the request, then the extension will set
+  // the ``name`` field specified header value to the upstream host address in the response anyway.
+  //
+  // Only one of ``override_mode_config`` or ``packages_mode_config`` should be set. This is the
+  // default if no any mode config is set.
+  OverrideModeConfig override_mode_config = 2;
+
+  // Set this to enable packages mode. In this mode, iff the header specified by the ``name`` field
+  // is present in the response (assume the ``name`` is set to ``session-header`` and original
+  // header value is ``xxxxxx``), then the upstream host address and value of ``name`` field
+  // specified header will be encoded in following format and the output will be used to update the
+  // ``name`` field specified header in the response:
+  //
+  // .. code-block:: none
+  //
+  //   session-header: "MS4yLjMuNDo4MDtvcmlnaW46eHh4eHh4Cg==" # base64("1.2.3.4:80;origin:xxxxxx")
+  //
+  // The ``origin`` part is used to store the original header value of ``name`` field specified
+  // header. And the ``origin`` part should always be last part of the new header value.
```

Comment:
I think the ordering requirement is unnecessary, and complicates the interfaces. Other than simplifying the code, what's the motivation for keeping it the last element? The main reason is that adding ";cluster" or any other future component may change the ordering.
Moreover, it might be that a concatenation of 2 "origins" (e.g., `1.2.3.4:80;origin:abc;origin:xyz` in case the upstream set `abc;origin:xyz` as the value of the header), may cause some ambiguity.
IT may be possible to use base64 again, and use a delimiter which will not be part of the value.

Reply:
Yeah. then we need to do additional base64, but I am ok to that if we think it's deserved

Reply:
I'm not sure I fully agree, and I think the current format adds a constraint on the allowed contents of the header value.
Say that the stateful-session-header name is `ssheader`, and that the upstream sends back the following `ssheader` value: `ssheader:abc;origin:xyz` (i.e., `ssheader` = `abc;origin:xyz`, which is a valid header value AFAIK).
This will be sent back to the downstream with the endpoint's address as follows `1.2.3.4:80;origin:ssheader:abc;origin:xyz`. Now when the stateful-session filter parses this value, it will take the last part and forward the value of `xyz`.

Reply:
I may be missing something, but was this addressed, or was there any followup to this comment?